Pipeline failed while executing TritonGPURemoveLayoutConversions: operand #0 does not dominate this use

### Describe the bug

Repro:
```py
import torch
import triton
import triton.language as tl

@triton.jit
def promote_to_tensor(x):
    return x + tl.zeros((1,), tl.int1)


@triton.jit
def is_floating(x):
    return promote_to_tensor(x).dtype.is_floating()


@triton.jit
def maximum(a, b):
    mask = a > b
    if is_floating(a):
        mask |= a != a
    return tl.where(mask, a, b)


@triton.jit
def maximum_with_index(a_value, a_index, b_value, b_index):
    mask = a_value > b_value
    equal = a_value == b_value
    if is_floating(a_value):
        a_isnan = a_value != a_value
        b_isnan = b_value != b_value
        mask |= a_isnan & (not b_isnan)
        equal |= a_isnan & b_isnan
    mask |= equal & (a_index < b_index)
    return tl.where(mask, a_value, b_value), tl.where(mask, a_index, b_index)


@triton.jit
def max2(a, dim):
    return tl.reduce(a, dim, maximum)


@triton.jit
def max_with_index(value, index, dim):
    return tl.reduce((value, index), dim, maximum_with_index)


libdevice = tl.extra.cuda.libdevice
tl_math = tl.math


@triton.jit
def triton_red_fused_add_div_eq_exp_ge_log_logical_and_masked_fill_max_mul_neg_pow_scatter_sub_unsqueeze_where_zeros_1(in_out_ptr1, in_out_ptr2, in_out_ptr4, in_out_ptr5, in_out_ptr6, in_ptr0, in_ptr1, in_ptr2, in_ptr3, in_ptr4, in_ptr5, in_ptr6, out_ptr8, ks0, ks1, ks2, ks3, ks4, xnumel, r0_numel, XBLOCK: tl.constexpr, R0_BLOCK: tl.constexpr):
    rnumel = r0_numel
    RBLOCK: tl.constexpr = R0_BLOCK
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    r0_base = tl.arange(0, R0_BLOCK)[None, :]
    rbase = r0_base
    x0 = xindex
    _tmp2 = tl.full([XBLOCK, R0_BLOCK], float("-inf"), tl.float32)
    _tmp4 = tl.full([XBLOCK, R0_BLOCK], float("-inf"), tl.float32)
    _tmp4_index = tl.full([XBLOCK, R0_BLOCK], 2147483647, tl.int32)
    for r0_offset in range(0, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_1 = r0_index
        tmp0 = tl.load(in_ptr0 + (r0_1 + x0 + ks0*r0_1 + ks1*r0_1 + ks2*r0_1 + ks3*r0_1), r0_mask & xmask, eviction_policy='evict_first', other=0.0).to(tl.float32)
        tmp1 = tl.broadcast_to(tmp0, [XBLOCK, R0_BLOCK])
        tmp3 = maximum(_tmp2, tmp1)
        _tmp2 = tl.where(r0_mask & xmask, tmp3, _tmp2)
        _tmp4_next, _tmp4_index_next = maximum_with_index(
            _tmp4, _tmp4_index, tmp1, rindex
        )
        _tmp4 = tl.where(r0_mask & xmask, _tmp4_next, _tmp4)
        _tmp4_index = tl.where(r0_mask & xmask, _tmp4_index_next, _tmp4_index)
    tmp2 = max2(_tmp2, 1)[:, None]
    tmp4_val, tmp4_idx = max_with_index(_tmp4, _tmp4_index, 1)
    tmp4 = tmp4_idx[:, None]
    tmp5 = tl.load(in_ptr1 + (x0), xmask, eviction_policy='evict_last').to(tl.float32)
    tmp15 = tl.load(in_ptr2 + (x0), xmask, eviction_policy='evict_last').to(tl.float32)
    tmp23 = tl.load(in_ptr3 + (x0), xmask, eviction_policy='evict_last').to(tl.float32)
    tmp33 = tl.load(in_ptr4 + (x0), xmask, eviction_policy='evict_last').to(tl.float32)
    tmp103 = tl.load(in_ptr5 + (x0), xmask, eviction_policy='evict_last').to(tl.float32)
    tmp116 = tl.load(in_ptr6 + (x0), xmask, eviction_policy='evict_last').to(tl.float32)
    tmp6 = tmp5.to(tl.float16)
    tmp7 = tmp6.to(tl.float32)
    tmp8 = tmp2.to(tl.float16)
    tmp9 = tmp8.to(tl.float32)
    tmp10 = libdevice.pow(tmp7, tmp9)
    tmp11 = tmp10.to(tl.float16)
    tmp12 = tmp11.to(tl.float32)
    tmp13 = tmp12.to(tl.float16)
    tmp14 = tmp13.to(tl.float32)
    tmp16 = tmp15.to(tl.float16)
    tmp17 = tmp16.to(tl.float32)
    tmp18 = libdevice.pow(tmp14, tmp17)
    tmp19 = tmp18.to(tl.float16)
    tmp20 = tmp19.to(tl.float32)
    tmp21 = tmp20.to(tl.float16)
    tmp22 = tmp21.to(tl.float32)
    tmp24 = tmp23.to(tl.float16)
    tmp25 = tmp24.to(tl.float32)
    tmp26 = libdevice.pow(tmp22, tmp25)
    tmp27 = tmp26.to(tl.float16)
    tmp28 = tmp27.to(tl.float32)
    tmp29 = tmp28.to(tl.float16)
    tmp30 = tmp29.to(tl.float32)
    tmp31 = 0.0
    tmp32 = tmp30 == tmp31
    tmp34 = tmp33.to(tl.float16)
    tmp35 = tmp34.to(tl.float32)
    tmp36 = libdevice.exp(tmp35)
    tmp37 = tmp36.to(tl.float16)
    tmp38 = tmp37.to(tl.float32)
    tmp39 = tmp38.to(tl.float16)
    tmp40 = tmp39.to(tl.float32)
    tmp41 = libdevice.pow(tmp30, tmp40)
    tmp42 = tmp41.to(tl.float16)
    tmp43 = tmp42.to(tl.float32)
    tmp44 = tmp43.to(tl.float16)
    tmp45 = tmp44.to(tl.float32)
    tmp46 = tl_math.log(tmp30)
    tmp47 = tmp46.to(tl.float16)
    tmp48 = tmp47.to(tl.float32)
    tmp49 = tmp48.to(tl.float16)
    tmp50 = tmp49.to(tl.float32)
    tmp51 = tmp45 * tmp50
    tmp52 = tmp51.to(tl.float16)
    tmp53 = tmp52.to(tl.float32)
    tmp54 = 1.0
    tmp55 = tmp40 - tmp54
    tmp56 = tmp55.to(tl.float16)
    tmp57 = tmp56.to(tl.float32)
    tmp58 = tmp57.to(tl.float16)
    tmp59 = tmp58.to(tl.float32)
    tmp60 = libdevice.pow(tmp30, tmp59)
    tmp61 = tmp60.to(tl.float16)
    tmp62 = tmp61.to(tl.float32)
    tmp63 = tl_math.log(tmp22)
    tmp64 = tmp63.to(tl.float16)
    tmp65 = tmp64.to(tl.float32)
    tmp66 = tmp65.to(tl.float16)
    tmp67 = tmp66.to(tl.float32)
    tmp68 = tmp30 * tmp67
    tmp69 = tmp68.to(tl.float16)
    tmp70 = tmp69.to(tl.float32)
    tmp71 = tmp22 == tmp31
    tmp72 = tmp25 >= tmp31
    tmp73 = tmp71 & tmp72
    tmp74 = tmp70.to(tl.float16)
    tmp75 = tmp74.to(tl.float32)
    tmp76 = tl.where(tmp73, tmp31, tmp75)
    tmp77 = tmp76.to(tl.float16)
    tmp78 = tmp77.to(tl.float32)
    tmp79 = tmp25 - tmp54
    tmp80 = tmp79.to(tl.float16)
    tmp81 = tmp80.to(tl.float32)
    tmp82 = tmp81.to(tl.float16)
    tmp83 = tmp82.to(tl.float32)
    tmp84 = libdevice.pow(tmp22, tmp83)
    tmp85 = tmp84.to(tl.float16)
    tmp86 = tmp85.to(tl.float32)
    tmp87 = tmp14 == tmp31
    tmp88 = tmp17 >= tmp31
    tmp89 = tmp87 & tmp88
    tmp90 = tl_math.log(tmp14)
    tmp91 = tmp90.to(tl.float16)
    tmp92 = tmp91.to(tl.float32)
    tmp93 = tmp92.to(tl.float16)
    tmp94 = tmp93.to(tl.float32)
    tmp95 = tmp22 * tmp94
    tmp96 = tmp95.to(tl.float16)
    tmp97 = tmp96.to(tl.float32)
    tmp98 = tmp97.to(tl.float16)
    tmp99 = tmp98.to(tl.float32)
    tmp100 = tl.where(tmp89, tmp31, tmp99)
    tmp101 = tmp100.to(tl.float16)
    tmp102 = tmp101.to(tl.float32)
    tmp104 = tmp103.to(tl.float16)
    tmp105 = tmp104.to(tl.float32)
    tmp106 = -tmp105
    tmp107 = tmp106.to(tl.float16)
    tmp108 = tmp107.to(tl.float32)
    tmp109 = tmp108.to(tl.float16)
    tmp110 = tmp109.to(tl.float32)
    tmp111 = (tmp45 / tmp45)
    tmp112 = tmp111.to(tl.float16)
    tmp113 = tmp112.to(tl.float32)
    tmp114 = tmp113.to(tl.float16)
    tmp115 = tmp114.to(tl.float32)
    tmp117 = tmp116.to(tl.float16)
    tmp118 = tmp117.to(tl.float32)
    tmp119 = (tmp115 / tmp118)
    tmp120 = tmp119.to(tl.float16)
    tmp121 = tmp120.to(tl.float32)
    tmp122 = tmp121.to(tl.float16)
    tmp123 = tmp122.to(tl.float32)
    tmp124 = (tmp123 / tmp118)
    tmp125 = tmp124.to(tl.float16)
    tmp126 = tmp125.to(tl.float32)
    tmp127 = tmp126.to(tl.float16)
    tmp128 = tmp127.to(tl.float32)
    tmp129 = (tmp128 / tmp118)
    tmp130 = tmp129.to(tl.float16)
    tmp131 = tmp130.to(tl.float32)
    tmp132 = tmp131.to(tl.float16)
    tmp133 = tmp132.to(tl.float32)
    tmp134 = tmp110 * tmp133
    tmp135 = tmp134.to(tl.float16)
    tmp136 = tmp135.to(tl.float32)
    tmp137 = (tmp105 / tmp118)
    tmp138 = tmp137.to(tl.float16)
    tmp139 = tmp138.to(tl.float32)
    tmp140 = tmp139.to(tl.float16)
    tmp141 = tmp140.to(tl.float32)
    tmp142 = -tmp141
    tmp143 = tmp142.to(tl.float16)
    tmp144 = tmp143.to(tl.float32)
    tmp145 = tmp144.to(tl.float16)
    tmp146 = tmp145.to(tl.float32)
    tmp147 = tmp146 * tmp128
    tmp148 = tmp147.to(tl.float16)
    tmp149 = tmp148.to(tl.float32)
    tmp150 = (tmp141 / tmp118)
    tmp151 = tmp150.to(tl.float16)
    tmp152 = tmp151.to(tl.float32)
    tmp153 = tmp152.to(tl.float16)
    tmp154 = tmp153.to(tl.float32)
    tmp155 = -tmp154
    tmp156 = tmp155.to(tl.float16)
    tmp157 = tmp156.to(tl.float32)
    tmp158 = tmp157.to(tl.float16)
    tmp159 = tmp158.to(tl.float32)
    tmp160 = (tmp115 / tmp45)
    tmp161 = tmp160.to(tl.float16)
    tmp162 = tmp161.to(tl.float32)
    tmp163 = tmp162.to(tl.float16)
    tmp164 = tmp163.to(tl.float32)
    tmp165 = tmp159 * tmp164
    tmp166 = tmp165.to(tl.float16)
    tmp167 = tmp166.to(tl.float32)
    tmp168 = (tmp154 / tmp45)
    tmp169 = tmp168.to(tl.float16)
    tmp170 = tmp169.to(tl.float32)
    tmp171 = tmp170.to(tl.float16)
    tmp172 = tmp171.to(tl.float32)
    tmp173 = tmp167.to(tl.float16)
    tmp174 = tmp173.to(tl.float32)
    tmp175 = tmp172 + tmp174
    tmp176 = tmp175.to(tl.float16)
    tmp177 = tmp176.to(tl.float32)
    tmp178 = tmp177.to(tl.float16)
    tmp179 = tmp178.to(tl.float32)
    tmp180 = tmp40 >= tmp31
    tmp181 = tmp32 & tmp180
    tmp182 = tmp53.to(tl.float16)
    tmp183 = tmp182.to(tl.float32)
    tmp184 = tl.where(tmp181, tmp31, tmp183)
    tmp185 = tmp184.to(tl.float16)
    tmp186 = tmp185.to(tl.float32)
    tmp187 = tmp186.to(tl.float16)
    tmp188 = tmp187.to(tl.float32)
    tmp189 = tmp179 * tmp188
    tmp190 = tmp189.to(tl.float16)
    tmp191 = tmp190.to(tl.float32)
    tmp192 = tmp62.to(tl.float16)
    tmp193 = tmp192.to(tl.float32)
    tmp194 = tmp40 * tmp193
    tmp195 = tmp194.to(tl.float16)
    tmp196 = tmp195.to(tl.float32)
    tmp197 = tmp196.to(tl.float16)
    tmp198 = tmp197.to(tl.float32)
    tmp199 = tmp179 * tmp198
    tmp200 = tmp199.to(tl.float16)
    tmp201 = tmp200.to(tl.float32)
    tmp202 = tmp40 == tmp31
    tmp203 = tmp201.to(tl.float16)
    tmp204 = tmp203.to(tl.float32)
    tmp205 = tl.where(tmp202, tmp31, tmp204)
    tmp206 = tmp205.to(tl.float16)
    tmp207 = tmp206.to(tl.float32)
    tmp208 = tmp207.to(tl.float16)
    tmp209 = tmp208.to(tl.float32)
    tmp210 = tmp86.to(tl.float16)
    tmp211 = tmp210.to(tl.float32)
    tmp212 = tmp25 * tmp211
    tmp213 = tmp212.to(tl.float16)
    tmp214 = tmp213.to(tl.float32)
    tmp215 = tmp214.to(tl.float16)
    tmp216 = tmp215.to(tl.float32)
    tmp217 = tmp209 * tmp216
    tmp218 = tmp217.to(tl.float16)
    tmp219 = tmp218.to(tl.float32)
    tmp220 = tmp25 == tmp31
    tmp221 = tmp219.to(tl.float16)
    tmp222 = tmp221.to(tl.float32)
    tmp223 = tl.where(tmp220, tmp31, tmp222)
    tmp224 = tmp223.to(tl.float16)
    tmp225 = tmp224.to(tl.float32)
    tmp226 = tmp225.to(tl.float16)
    tmp227 = tmp226.to(tl.float32)
    tmp228 = tmp102.to(tl.float16)
    tmp229 = tmp228.to(tl.float32)
    tmp230 = tmp227 * tmp229
    tmp231 = tmp230.to(tl.float16)
    tmp232 = tmp231.to(tl.float32)
    tmp233 = tmp17 - tmp54
    tmp234 = tmp233.to(tl.float16)
    tmp235 = tmp234.to(tl.float32)
    tmp236 = tmp235.to(tl.float16)
    tmp237 = tmp236.to(tl.float32)
    tmp238 = libdevice.pow(tmp14, tmp237)
    tmp239 = tmp238.to(tl.float16)
    tmp240 = tmp239.to(tl.float32)
    tmp241 = tmp240.to(tl.float16)
    tmp242 = tmp241.to(tl.float32)
    tmp243 = tmp17 * tmp242
    tmp244 = tmp243.to(tl.float16)
    tmp245 = tmp244.to(tl.float32)
    tmp246 = tmp245.to(tl.float16)
    tmp247 = tmp246.to(tl.float32)
    tmp248 = tmp227 * tmp247
    tmp249 = tmp248.to(tl.float16)
    tmp250 = tmp249.to(tl.float32)
    tmp251 = tmp17 == tmp31
    tmp252 = tmp250.to(tl.float16)
    tmp253 = tmp252.to(tl.float32)
    tmp254 = tl.where(tmp251, tmp31, tmp253)
    tmp255 = tmp254.to(tl.float16)
    tmp256 = tmp255.to(tl.float32)
    tmp257 = tmp256.to(tl.float16)
    tmp258 = tmp257.to(tl.float32)
    tmp259 = tmp9 - tmp54
    tmp260 = tmp259.to(tl.float16)
    tmp261 = tmp260.to(tl.float32)
    tmp262 = tmp261.to(tl.float16)
    tmp263 = tmp262.to(tl.float32)
    tmp264 = libdevice.pow(tmp7, tmp263)
    tmp265 = tmp264.to(tl.float16)
    tmp266 = tmp265.to(tl.float32)
    tmp267 = tmp266.to(tl.float16)
    tmp268 = tmp267.to(tl.float32)
    tmp269 = tmp9 * tmp268
    tmp270 = tmp269.to(tl.float16)
    tmp271 = tmp270.to(tl.float32)
    tmp272 = tmp271.to(tl.float16)
    tmp273 = tmp272.to(tl.float32)
    tmp274 = tmp258 * tmp273
    tmp275 = tmp274.to(tl.float16)
    tmp276 = tmp275.to(tl.float32)
    tmp277 = tmp7 == tmp31
    tmp278 = tmp9 >= tmp31
    tmp279 = tmp277 & tmp278
    tmp280 = tl_math.log(tmp7)
    tmp281 = tmp280.to(tl.float16)
    tmp282 = tmp281.to(tl.float32)
    tmp283 = tmp282.to(tl.float16)
    tmp284 = tmp283.to(tl.float32)
    tmp285 = tmp14 * tmp284
    tmp286 = tmp285.to(tl.float16)
    tmp287 = tmp286.to(tl.float32)
    tmp288 = tmp287.to(tl.float16)
    tmp289 = tmp288.to(tl.float32)
    tmp290 = tl.where(tmp279, tmp31, tmp289)
    tmp291 = tmp290.to(tl.float16)
    tmp292 = tmp291.to(tl.float32)
    tmp293 = tmp9 == tmp31
    tmp294 = tmp276.to(tl.float16)
    tmp295 = tmp294.to(tl.float32)
    tmp296 = tl.where(tmp293, tmp31, tmp295)
    tmp297 = tmp296.to(tl.float16)
    tmp298 = tmp297.to(tl.float32)
    tl.device_assert(((0 <= tmp4) & (tmp4 < ks4)) | ~(xmask), "index out of bounds: 0 <= tmp4 < ks4")
    tmp300 = tmp292.to(tl.float16)
    tmp301 = tmp300.to(tl.float32)
    tmp302 = tmp258 * tmp301
    tmp303 = tmp302.to(tl.float16)
    tmp304 = tmp303.to(tl.float32)
    tmp305 = tmp191.to(tl.float16)
    tmp306 = tmp305.to(tl.float32)
    tmp307 = tmp306 * tmp40
    tmp308 = tmp307.to(tl.float16)
    tmp309 = tmp308.to(tl.float32)
    tmp310 = tmp78.to(tl.float16)
    tmp311 = tmp310.to(tl.float32)
    tmp312 = tmp209 * tmp311
    tmp313 = tmp312.to(tl.float16)
    tmp314 = tmp313.to(tl.float32)
    tmp315 = tmp314.to(tl.float16)
    tmp316 = tmp315.to(tl.float32)
    tmp317 = tmp232.to(tl.float16)
    tmp318 = tmp317.to(tl.float32)
    tmp319 = tmp316 + tmp318
    tmp320 = tmp319.to(tl.float16)
    tmp321 = tmp320.to(tl.float32)
    tmp322 = tmp136.to(tl.float16)
    tmp323 = tmp322.to(tl.float32)
    tmp324 = tmp149.to(tl.float16)
    tmp325 = tmp324.to(tl.float32)
    tmp326 = tmp323 + tmp325
    tmp327 = tmp326.to(tl.float16)
    tmp328 = tmp327.to(tl.float32)
    tl.debug_barrier()
    tl.store(in_out_ptr4 + (x0), tmp232, xmask)
    tl.debug_barrier()
    tl.store(in_out_ptr5 + (x0), tmp298, xmask)
    tl.store(out_ptr8 + (tmp4 + x0 + ks0*tmp4 + ks1*tmp4 + ks2*tmp4 + ks3*tmp4), tmp304, xmask)
    tl.debug_barrier()
    tl.store(in_out_ptr1 + (x0), tmp309, xmask)
    tl.debug_barrier()
    tl.store(in_out_ptr2 + (x0), tmp321, xmask)
    tl.debug_barrier()
    tl.store(in_out_ptr6 + (x0), tmp328, xmask)


def main():
    device = torch.device("cuda")
    shape_1d = (241,)
    def rand1d():
        return torch.rand(shape_1d, device=device, dtype=torch.float16)

    in_out_ptr1 = rand1d()
    in_out_ptr2 = rand1d()
    in_out_ptr4 = rand1d()
    in_out_ptr5 = rand1d()
    in_out_ptr6 = rand1d()
    in_ptr0 = torch.rand((3, 241), device=device, dtype=torch.float16)
    in_ptr1 = rand1d()
    in_ptr2 = rand1d()
    in_ptr3 = rand1d()
    in_ptr4 = rand1d()
    in_ptr5 = rand1d()
    in_ptr6 = rand1d()
    out_ptr8 = torch.empty((3, 241), device=device, dtype=torch.float16)

    ks0, ks1, ks2, ks3, ks4 = 26, 111, 13, 90, 3
    xnumel = 241
    r0_numel = 3

    grid = lambda meta: (triton.cdiv(xnumel, meta["XBLOCK"]),)

    triton_red_fused_add_div_eq_exp_ge_log_logical_and_masked_fill_max_mul_neg_pow_scatter_sub_unsqueeze_where_zeros_1[
        grid
    ](
        in_out_ptr1,
        in_out_ptr2,
        in_out_ptr4,
        in_out_ptr5,
        in_out_ptr6,
        in_ptr0,
        in_ptr1,
        in_ptr2,
        in_ptr3,
        in_ptr4,
        in_ptr5,
        in_ptr6,
        out_ptr8,
        ks0,
        ks1,
        ks2,
        ks3,
        ks4,
        xnumel,
        r0_numel,
        XBLOCK=8,
        R0_BLOCK=4,
    )


if __name__ == "__main__":
    main()
```

Output:
```
/home/jansel/pytorch/triton_issue163567_min.py:394:23: error: operand #0 does not dominate this use
    tmp318 = tmp317.to(tl.float32)
                      ^
/home/jansel/pytorch/triton_issue163567_min.py:406:41: note: operand defined here (op in the same block)
    tl.store(in_out_ptr4 + (x0), tmp232, xmask)
                                        ^
module {
  tt.func public @triton_red_fused_add_div_eq_exp_ge_log_logical_and_masked_fill_max_mul_neg_pow_scatter_sub_unsqueeze_where_zeros_1(%arg0: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg1: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg2: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg3: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg4: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg5: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg6: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg7: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg8: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg9: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg10: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg11: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg12: !tt.ptr<f16> {tt.divisibility = 16 : i32}, %arg13: i32, %arg14: i32, %arg15: i32, %arg16: i32, %arg17: i32, %arg18: i32, %arg19: i32) attributes {noinline = false} {
    %true = arith.constant true
    %cst = arith.constant dense<true> : tensor<8x4xi1>
    %cst_0 = arith.constant dense<0.000000e+00> : tensor<8x4xf16>
    %c4_i32 = arith.constant 4 : i32
    %c0_i32 = arith.constant 0 : i32
    %cst_1 = arith.constant dense<1.000000e+00> : tensor<8x1xf32>
    %cst_2 = arith.constant dense<0.000000e+00> : tensor<8x1xf32>
    %cst_3 = arith.constant dense<2147483647> : tensor<8x4xi32>
    %cst_4 = arith.constant dense<0xFF800000> : tensor<8x4xf32>
    %c8_i32 = arith.constant 8 : i32
    %0 = tt.get_program_id x : i32
    %1 = arith.muli %0, %c8_i32 : i32
    %2 = tt.make_range {end = 8 : i32, start = 0 : i32} : tensor<8xi32>
    %3 = tt.expand_dims %2 {axis = 1 : i32} : tensor<8xi32> -> tensor<8x1xi32>
    %4 = tt.splat %1 : i32 -> tensor<8x1xi32>
    %5 = arith.addi %4, %3 : tensor<8x1xi32>
    %6 = tt.splat %arg18 : i32 -> tensor<8x1xi32>
    %7 = arith.cmpi slt, %5, %6 : tensor<8x1xi32>
    %8 = tt.make_range {end = 4 : i32, start = 0 : i32} : tensor<4xi32>
    %9 = tt.expand_dims %8 {axis = 0 : i32} : tensor<4xi32> -> tensor<1x4xi32>
    %10:3 = scf.for %arg20 = %c0_i32 to %arg19 step %c4_i32 iter_args(%arg21 = %cst_4, %arg22 = %cst_4, %arg23 = %cst_3) -> (tensor<8x4xf32>, tensor<8x4xf32>, tensor<8x4xi32>)  : i32 {
      %254 = tt.splat %arg20 : i32 -> tensor<1x4xi32>
      %255 = arith.addi %254, %9 : tensor<1x4xi32>
      %256 = tt.splat %arg19 : i32 -> tensor<1x4xi32>
      %257 = arith.cmpi slt, %255, %256 : tensor<1x4xi32>
      %258 = tt.broadcast %255 : tensor<1x4xi32> -> tensor<8x4xi32>
      %259 = tt.broadcast %5 : tensor<8x1xi32> -> tensor<8x4xi32>
      %260 = arith.addi %258, %259 : tensor<8x4xi32>
      %261 = tt.splat %arg13 : i32 -> tensor<1x4xi32>
      %262 = arith.muli %261, %255 : tensor<1x4xi32>
      %263 = tt.broadcast %262 : tensor<1x4xi32> -> tensor<8x4xi32>
      %264 = arith.addi %260, %263 : tensor<8x4xi32>
      %265 = tt.splat %arg14 : i32 -> tensor<1x4xi32>
      %266 = arith.muli %265, %255 : tensor<1x4xi32>
      %267 = tt.broadcast %266 : tensor<1x4xi32> -> tensor<8x4xi32>
      %268 = arith.addi %264, %267 : tensor<8x4xi32>
      %269 = tt.splat %arg15 : i32 -> tensor<1x4xi32>
      %270 = arith.muli %269, %255 : tensor<1x4xi32>
      %271 = tt.broadcast %270 : tensor<1x4xi32> -> tensor<8x4xi32>
      %272 = arith.addi %268, %271 : tensor<8x4xi32>
      %273 = tt.splat %arg16 : i32 -> tensor<1x4xi32>
      %274 = arith.muli %273, %255 : tensor<1x4xi32>
      %275 = tt.broadcast %274 : tensor<1x4xi32> -> tensor<8x4xi32>
      %276 = arith.addi %272, %275 : tensor<8x4xi32>
      %277 = tt.splat %arg5 : !tt.ptr<f16> -> tensor<8x4x!tt.ptr<f16>>
      %278 = tt.addptr %277, %276 : tensor<8x4x!tt.ptr<f16>>, tensor<8x4xi32>
      %279 = tt.broadcast %257 : tensor<1x4xi1> -> tensor<8x4xi1>
      %280 = tt.broadcast %7 : tensor<8x1xi1> -> tensor<8x4xi1>
      %281 = arith.andi %279, %280 : tensor<8x4xi1>
      %282 = tt.load %278, %281, %cst_0 evictionPolicy = evict_first : tensor<8x4x!tt.ptr<f16>>
      %283 = arith.extf %282 : tensor<8x4xf16> to tensor<8x4xf32>
      %284 = arith.cmpf ogt, %arg21, %283 : tensor<8x4xf32>
      %285 = arith.cmpf une, %arg21, %arg21 : tensor<8x4xf32>
      %286 = arith.ori %284, %285 : tensor<8x4xi1>
      %287 = arith.select %286, %arg21, %283 : tensor<8x4xi1>, tensor<8x4xf32>
      %288 = arith.select %281, %287, %arg21 : tensor<8x4xi1>, tensor<8x4xf32>
      %289 = arith.cmpf ogt, %arg22, %283 : tensor<8x4xf32>
      %290 = arith.cmpf oeq, %arg22, %283 : tensor<8x4xf32>
      %291 = arith.cmpf une, %arg22, %arg22 : tensor<8x4xf32>
      %292 = arith.cmpf une, %283, %283 : tensor<8x4xf32>
      %293 = arith.xori %292, %cst : tensor<8x4xi1>
      %294 = arith.andi %291, %293 : tensor<8x4xi1>
      %295 = arith.ori %289, %294 : tensor<8x4xi1>
      %296 = arith.andi %291, %292 : tensor<8x4xi1>
      %297 = arith.ori %290, %296 : tensor<8x4xi1>
      %298 = arith.cmpi slt, %arg23, %258 : tensor<8x4xi32>
      %299 = arith.andi %297, %298 : tensor<8x4xi1>
      %300 = arith.ori %295, %299 : tensor<8x4xi1>
      %301 = arith.select %300, %arg22, %283 : tensor<8x4xi1>, tensor<8x4xf32>
      %302 = arith.select %300, %arg23, %258 : tensor<8x4xi1>, tensor<8x4xi32>
      %303 = arith.select %281, %301, %arg22 : tensor<8x4xi1>, tensor<8x4xf32>
      %304 = arith.select %281, %302, %arg23 : tensor<8x4xi1>, tensor<8x4xi32>
      scf.yield %288, %303, %304 : tensor<8x4xf32>, tensor<8x4xf32>, tensor<8x4xi32>
    }
    %11 = "tt.reduce"(%10#0) <{axis = 1 : i32}> ({
    ^bb0(%arg20: f32, %arg21: f32):
      %254 = arith.cmpf ogt, %arg20, %arg21 : f32
      %255 = arith.cmpf une, %arg20, %arg20 : f32
      %256 = arith.ori %254, %255 : i1
      %257 = arith.select %256, %arg20, %arg21 : f32
      tt.reduce.return %257 : f32
    }) : (tensor<8x4xf32>) -> tensor<8xf32>
    %12 = tt.expand_dims %11 {axis = 1 : i32} : tensor<8xf32> -> tensor<8x1xf32>
    %13:2 = "tt.reduce"(%10#1, %10#2) <{axis = 1 : i32}> ({
    ^bb0(%arg20: f32, %arg21: i32, %arg22: f32, %arg23: i32):
      %254 = arith.cmpf ogt, %arg20, %arg22 : f32
      %255 = arith.cmpf oeq, %arg20, %arg22 : f32
      %256 = arith.cmpf une, %arg20, %arg20 : f32
      %257 = arith.cmpf une, %arg22, %arg22 : f32
      %258 = arith.xori %257, %true : i1
      %259 = arith.andi %256, %258 : i1
      %260 = arith.ori %254, %259 : i1
      %261 = arith.andi %256, %257 : i1
      %262 = arith.ori %255, %261 : i1
      %263 = arith.cmpi slt, %arg21, %arg23 : i32
      %264 = arith.andi %262, %263 : i1
      %265 = arith.ori %260, %264 : i1
      %266 = arith.select %265, %arg20, %arg22 : f32
      %267 = arith.select %265, %arg21, %arg23 : i32
      tt.reduce.return %266, %267 : f32, i32
    }) : (tensor<8x4xf32>, tensor<8x4xi32>) -> (tensor<8xf32>, tensor<8xi32>)
    %14 = tt.expand_dims %13#1 {axis = 1 : i32} : tensor<8xi32> -> tensor<8x1xi32>
    %15 = tt.splat %arg6 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %16 = tt.addptr %15, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    %17 = tt.load %16, %7 evictionPolicy = evict_last : tensor<8x1x!tt.ptr<f16>>
    %18 = tt.splat %arg7 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %19 = tt.addptr %18, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    %20 = tt.load %19, %7 evictionPolicy = evict_last : tensor<8x1x!tt.ptr<f16>>
    %21 = tt.splat %arg8 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %22 = tt.addptr %21, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    %23 = tt.load %22, %7 evictionPolicy = evict_last : tensor<8x1x!tt.ptr<f16>>
    %24 = tt.splat %arg9 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %25 = tt.addptr %24, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    %26 = tt.load %25, %7 evictionPolicy = evict_last : tensor<8x1x!tt.ptr<f16>>
    %27 = tt.splat %arg10 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %28 = tt.addptr %27, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    %29 = tt.load %28, %7 evictionPolicy = evict_last : tensor<8x1x!tt.ptr<f16>>
    %30 = tt.splat %arg11 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %31 = tt.addptr %30, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    %32 = tt.load %31, %7 evictionPolicy = evict_last : tensor<8x1x!tt.ptr<f16>>
    %33 = arith.extf %17 : tensor<8x1xf16> to tensor<8x1xf32>
    %34 = arith.truncf %12 : tensor<8x1xf32> to tensor<8x1xf16>
    %35 = arith.extf %34 : tensor<8x1xf16> to tensor<8x1xf32>
    %36 = tt.extern_elementwise %33, %35 {libname = "", libpath = "", pure = true, symbol = "__nv_powf"} : (tensor<8x1xf32>, tensor<8x1xf32>) -> tensor<8x1xf32>
    %37 = arith.truncf %36 : tensor<8x1xf32> to tensor<8x1xf16>
    %38 = arith.extf %37 : tensor<8x1xf16> to tensor<8x1xf32>
    %39 = arith.extf %20 : tensor<8x1xf16> to tensor<8x1xf32>
    %40 = tt.extern_elementwise %38, %39 {libname = "", libpath = "", pure = true, symbol = "__nv_powf"} : (tensor<8x1xf32>, tensor<8x1xf32>) -> tensor<8x1xf32>
    %41 = arith.truncf %40 : tensor<8x1xf32> to tensor<8x1xf16>
    %42 = arith.extf %41 : tensor<8x1xf16> to tensor<8x1xf32>
    %43 = arith.extf %23 : tensor<8x1xf16> to tensor<8x1xf32>
    %44 = tt.extern_elementwise %42, %43 {libname = "", libpath = "", pure = true, symbol = "__nv_powf"} : (tensor<8x1xf32>, tensor<8x1xf32>) -> tensor<8x1xf32>
    %45 = arith.truncf %44 : tensor<8x1xf32> to tensor<8x1xf16>
    %46 = arith.extf %45 : tensor<8x1xf16> to tensor<8x1xf32>
    %47 = arith.cmpf oeq, %46, %cst_2 : tensor<8x1xf32>
    %48 = arith.extf %26 : tensor<8x1xf16> to tensor<8x1xf32>
    %49 = tt.extern_elementwise %48 {libname = "", libpath = "", pure = true, symbol = "__nv_expf"} : (tensor<8x1xf32>) -> tensor<8x1xf32>
    %50 = arith.truncf %49 : tensor<8x1xf32> to tensor<8x1xf16>
    %51 = arith.extf %50 : tensor<8x1xf16> to tensor<8x1xf32>
    %52 = tt.extern_elementwise %46, %51 {libname = "", libpath = "", pure = true, symbol = "__nv_powf"} : (tensor<8x1xf32>, tensor<8x1xf32>) -> tensor<8x1xf32>
    %53 = arith.truncf %52 : tensor<8x1xf32> to tensor<8x1xf16>
    %54 = arith.extf %53 : tensor<8x1xf16> to tensor<8x1xf32>
    %55 = math.log %46 : tensor<8x1xf32>
    %56 = arith.truncf %55 : tensor<8x1xf32> to tensor<8x1xf16>
    %57 = arith.extf %56 : tensor<8x1xf16> to tensor<8x1xf32>
    %58 = arith.mulf %54, %57 : tensor<8x1xf32>
    %59 = arith.truncf %58 : tensor<8x1xf32> to tensor<8x1xf16>
    %60 = arith.subf %51, %cst_1 : tensor<8x1xf32>
    %61 = arith.truncf %60 : tensor<8x1xf32> to tensor<8x1xf16>
    %62 = arith.extf %61 : tensor<8x1xf16> to tensor<8x1xf32>
    %63 = tt.extern_elementwise %46, %62 {libname = "", libpath = "", pure = true, symbol = "__nv_powf"} : (tensor<8x1xf32>, tensor<8x1xf32>) -> tensor<8x1xf32>
    %64 = arith.truncf %63 : tensor<8x1xf32> to tensor<8x1xf16>
    %65 = math.log %42 : tensor<8x1xf32>
    %66 = arith.truncf %65 : tensor<8x1xf32> to tensor<8x1xf16>
    %67 = arith.extf %66 : tensor<8x1xf16> to tensor<8x1xf32>
    %68 = arith.mulf %46, %67 : tensor<8x1xf32>
    %69 = arith.truncf %68 : tensor<8x1xf32> to tensor<8x1xf16>
    %70 = arith.cmpf oeq, %42, %cst_2 : tensor<8x1xf32>
    %71 = arith.cmpf oge, %43, %cst_2 : tensor<8x1xf32>
    %72 = arith.andi %70, %71 : tensor<8x1xi1>
    %73 = arith.extf %69 : tensor<8x1xf16> to tensor<8x1xf32>
    %74 = arith.select %72, %cst_2, %73 : tensor<8x1xi1>, tensor<8x1xf32>
    %75 = arith.truncf %74 : tensor<8x1xf32> to tensor<8x1xf16>
    %76 = arith.subf %43, %cst_1 : tensor<8x1xf32>
    %77 = arith.truncf %76 : tensor<8x1xf32> to tensor<8x1xf16>
    %78 = arith.extf %77 : tensor<8x1xf16> to tensor<8x1xf32>
    %79 = tt.extern_elementwise %42, %78 {libname = "", libpath = "", pure = true, symbol = "__nv_powf"} : (tensor<8x1xf32>, tensor<8x1xf32>) -> tensor<8x1xf32>
    %80 = arith.truncf %79 : tensor<8x1xf32> to tensor<8x1xf16>
    %81 = arith.cmpf oeq, %38, %cst_2 : tensor<8x1xf32>
    %82 = arith.cmpf oge, %39, %cst_2 : tensor<8x1xf32>
    %83 = arith.andi %81, %82 : tensor<8x1xi1>
    %84 = math.log %38 : tensor<8x1xf32>
    %85 = arith.truncf %84 : tensor<8x1xf32> to tensor<8x1xf16>
    %86 = arith.extf %85 : tensor<8x1xf16> to tensor<8x1xf32>
    %87 = arith.mulf %42, %86 : tensor<8x1xf32>
    %88 = arith.truncf %87 : tensor<8x1xf32> to tensor<8x1xf16>
    %89 = arith.extf %88 : tensor<8x1xf16> to tensor<8x1xf32>
    %90 = arith.select %83, %cst_2, %89 : tensor<8x1xi1>, tensor<8x1xf32>
    %91 = arith.truncf %90 : tensor<8x1xf32> to tensor<8x1xf16>
    %92 = arith.extf %29 : tensor<8x1xf16> to tensor<8x1xf32>
    %93 = arith.subf %cst_2, %92 : tensor<8x1xf32>
    %94 = arith.truncf %93 : tensor<8x1xf32> to tensor<8x1xf16>
    %95 = arith.extf %94 : tensor<8x1xf16> to tensor<8x1xf32>
    %96 = arith.divf %54, %54 : tensor<8x1xf32>
    %97 = arith.truncf %96 : tensor<8x1xf32> to tensor<8x1xf16>
    %98 = arith.extf %97 : tensor<8x1xf16> to tensor<8x1xf32>
    %99 = arith.extf %32 : tensor<8x1xf16> to tensor<8x1xf32>
    %100 = arith.divf %98, %99 : tensor<8x1xf32>
    %101 = arith.truncf %100 : tensor<8x1xf32> to tensor<8x1xf16>
    %102 = arith.extf %101 : tensor<8x1xf16> to tensor<8x1xf32>
    %103 = arith.divf %102, %99 : tensor<8x1xf32>
    %104 = arith.truncf %103 : tensor<8x1xf32> to tensor<8x1xf16>
    %105 = arith.extf %104 : tensor<8x1xf16> to tensor<8x1xf32>
    %106 = arith.divf %105, %99 : tensor<8x1xf32>
    %107 = arith.truncf %106 : tensor<8x1xf32> to tensor<8x1xf16>
    %108 = arith.extf %107 : tensor<8x1xf16> to tensor<8x1xf32>
    %109 = arith.mulf %95, %108 : tensor<8x1xf32>
    %110 = arith.truncf %109 : tensor<8x1xf32> to tensor<8x1xf16>
    %111 = arith.divf %92, %99 : tensor<8x1xf32>
    %112 = arith.truncf %111 : tensor<8x1xf32> to tensor<8x1xf16>
    %113 = arith.extf %112 : tensor<8x1xf16> to tensor<8x1xf32>
    %114 = arith.subf %cst_2, %113 : tensor<8x1xf32>
    %115 = arith.truncf %114 : tensor<8x1xf32> to tensor<8x1xf16>
    %116 = arith.extf %115 : tensor<8x1xf16> to tensor<8x1xf32>
    %117 = arith.mulf %116, %105 : tensor<8x1xf32>
    %118 = arith.truncf %117 : tensor<8x1xf32> to tensor<8x1xf16>
    %119 = arith.divf %113, %99 : tensor<8x1xf32>
    %120 = arith.truncf %119 : tensor<8x1xf32> to tensor<8x1xf16>
    %121 = arith.extf %120 : tensor<8x1xf16> to tensor<8x1xf32>
    %122 = arith.subf %cst_2, %121 : tensor<8x1xf32>
    %123 = arith.truncf %122 : tensor<8x1xf32> to tensor<8x1xf16>
    %124 = arith.extf %123 : tensor<8x1xf16> to tensor<8x1xf32>
    %125 = arith.divf %98, %54 : tensor<8x1xf32>
    %126 = arith.truncf %125 : tensor<8x1xf32> to tensor<8x1xf16>
    %127 = arith.extf %126 : tensor<8x1xf16> to tensor<8x1xf32>
    %128 = arith.mulf %124, %127 : tensor<8x1xf32>
    %129 = arith.truncf %128 : tensor<8x1xf32> to tensor<8x1xf16>
    %130 = arith.divf %121, %54 : tensor<8x1xf32>
    %131 = arith.truncf %130 : tensor<8x1xf32> to tensor<8x1xf16>
    %132 = arith.extf %131 : tensor<8x1xf16> to tensor<8x1xf32>
    %133 = arith.extf %129 : tensor<8x1xf16> to tensor<8x1xf32>
    %134 = arith.addf %132, %133 : tensor<8x1xf32>
    %135 = arith.truncf %134 : tensor<8x1xf32> to tensor<8x1xf16>
    %136 = arith.extf %135 : tensor<8x1xf16> to tensor<8x1xf32>
    %137 = arith.cmpf oge, %51, %cst_2 : tensor<8x1xf32>
    %138 = arith.andi %47, %137 : tensor<8x1xi1>
    %139 = arith.extf %59 : tensor<8x1xf16> to tensor<8x1xf32>
    %140 = arith.select %138, %cst_2, %139 : tensor<8x1xi1>, tensor<8x1xf32>
    %141 = arith.truncf %140 : tensor<8x1xf32> to tensor<8x1xf16>
    %142 = arith.extf %141 : tensor<8x1xf16> to tensor<8x1xf32>
    %143 = arith.mulf %136, %142 : tensor<8x1xf32>
    %144 = arith.truncf %143 : tensor<8x1xf32> to tensor<8x1xf16>
    %145 = arith.extf %64 : tensor<8x1xf16> to tensor<8x1xf32>
    %146 = arith.mulf %51, %145 : tensor<8x1xf32>
    %147 = arith.truncf %146 : tensor<8x1xf32> to tensor<8x1xf16>
    %148 = arith.extf %147 : tensor<8x1xf16> to tensor<8x1xf32>
    %149 = arith.mulf %136, %148 : tensor<8x1xf32>
    %150 = arith.truncf %149 : tensor<8x1xf32> to tensor<8x1xf16>
    %151 = arith.cmpf oeq, %51, %cst_2 : tensor<8x1xf32>
    %152 = arith.extf %150 : tensor<8x1xf16> to tensor<8x1xf32>
    %153 = arith.select %151, %cst_2, %152 : tensor<8x1xi1>, tensor<8x1xf32>
    %154 = arith.truncf %153 : tensor<8x1xf32> to tensor<8x1xf16>
    %155 = arith.extf %154 : tensor<8x1xf16> to tensor<8x1xf32>
    %156 = arith.extf %80 : tensor<8x1xf16> to tensor<8x1xf32>
    %157 = arith.mulf %43, %156 : tensor<8x1xf32>
    %158 = arith.truncf %157 : tensor<8x1xf32> to tensor<8x1xf16>
    %159 = arith.extf %158 : tensor<8x1xf16> to tensor<8x1xf32>
    %160 = arith.mulf %155, %159 : tensor<8x1xf32>
    %161 = arith.truncf %160 : tensor<8x1xf32> to tensor<8x1xf16>
    %162 = arith.cmpf oeq, %43, %cst_2 : tensor<8x1xf32>
    %163 = arith.extf %161 : tensor<8x1xf16> to tensor<8x1xf32>
    %164 = arith.select %162, %cst_2, %163 : tensor<8x1xi1>, tensor<8x1xf32>
    %165 = arith.truncf %164 : tensor<8x1xf32> to tensor<8x1xf16>
    %166 = arith.extf %165 : tensor<8x1xf16> to tensor<8x1xf32>
    %167 = arith.extf %91 : tensor<8x1xf16> to tensor<8x1xf32>
    %168 = arith.mulf %166, %167 : tensor<8x1xf32>
    %169 = arith.truncf %168 : tensor<8x1xf32> to tensor<8x1xf16>
    %170 = arith.subf %39, %cst_1 : tensor<8x1xf32>
    %171 = arith.truncf %170 : tensor<8x1xf32> to tensor<8x1xf16>
    %172 = arith.extf %171 : tensor<8x1xf16> to tensor<8x1xf32>
    %173 = tt.extern_elementwise %38, %172 {libname = "", libpath = "", pure = true, symbol = "__nv_powf"} : (tensor<8x1xf32>, tensor<8x1xf32>) -> tensor<8x1xf32>
    %174 = arith.truncf %173 : tensor<8x1xf32> to tensor<8x1xf16>
    %175 = arith.extf %174 : tensor<8x1xf16> to tensor<8x1xf32>
    %176 = arith.mulf %39, %175 : tensor<8x1xf32>
    %177 = arith.truncf %176 : tensor<8x1xf32> to tensor<8x1xf16>
    %178 = arith.extf %177 : tensor<8x1xf16> to tensor<8x1xf32>
    %179 = arith.mulf %166, %178 : tensor<8x1xf32>
    %180 = arith.truncf %179 : tensor<8x1xf32> to tensor<8x1xf16>
    %181 = arith.cmpf oeq, %39, %cst_2 : tensor<8x1xf32>
    %182 = arith.extf %180 : tensor<8x1xf16> to tensor<8x1xf32>
    %183 = arith.select %181, %cst_2, %182 : tensor<8x1xi1>, tensor<8x1xf32>
    %184 = arith.truncf %183 : tensor<8x1xf32> to tensor<8x1xf16>
    %185 = arith.extf %184 : tensor<8x1xf16> to tensor<8x1xf32>
    %186 = arith.subf %35, %cst_1 : tensor<8x1xf32>
    %187 = arith.truncf %186 : tensor<8x1xf32> to tensor<8x1xf16>
    %188 = arith.extf %187 : tensor<8x1xf16> to tensor<8x1xf32>
    %189 = tt.extern_elementwise %33, %188 {libname = "", libpath = "", pure = true, symbol = "__nv_powf"} : (tensor<8x1xf32>, tensor<8x1xf32>) -> tensor<8x1xf32>
    %190 = arith.truncf %189 : tensor<8x1xf32> to tensor<8x1xf16>
    %191 = arith.extf %190 : tensor<8x1xf16> to tensor<8x1xf32>
    %192 = arith.mulf %35, %191 : tensor<8x1xf32>
    %193 = arith.truncf %192 : tensor<8x1xf32> to tensor<8x1xf16>
    %194 = arith.extf %193 : tensor<8x1xf16> to tensor<8x1xf32>
    %195 = arith.mulf %185, %194 : tensor<8x1xf32>
    %196 = arith.truncf %195 : tensor<8x1xf32> to tensor<8x1xf16>
    %197 = arith.cmpf oeq, %33, %cst_2 : tensor<8x1xf32>
    %198 = arith.cmpf oge, %35, %cst_2 : tensor<8x1xf32>
    %199 = arith.andi %197, %198 : tensor<8x1xi1>
    %200 = math.log %33 : tensor<8x1xf32>
    %201 = arith.truncf %200 : tensor<8x1xf32> to tensor<8x1xf16>
    %202 = arith.extf %201 : tensor<8x1xf16> to tensor<8x1xf32>
    %203 = arith.mulf %38, %202 : tensor<8x1xf32>
    %204 = arith.truncf %203 : tensor<8x1xf32> to tensor<8x1xf16>
    %205 = arith.extf %204 : tensor<8x1xf16> to tensor<8x1xf32>
    %206 = arith.select %199, %cst_2, %205 : tensor<8x1xi1>, tensor<8x1xf32>
    %207 = arith.truncf %206 : tensor<8x1xf32> to tensor<8x1xf16>
    %208 = arith.cmpf oeq, %35, %cst_2 : tensor<8x1xf32>
    %209 = arith.extf %196 : tensor<8x1xf16> to tensor<8x1xf32>
    %210 = arith.select %208, %cst_2, %209 : tensor<8x1xi1>, tensor<8x1xf32>
    %211 = arith.truncf %210 : tensor<8x1xf32> to tensor<8x1xf16>
    %212 = arith.extf %207 : tensor<8x1xf16> to tensor<8x1xf32>
    %213 = arith.mulf %185, %212 : tensor<8x1xf32>
    %214 = arith.truncf %213 : tensor<8x1xf32> to tensor<8x1xf16>
    %215 = arith.extf %144 : tensor<8x1xf16> to tensor<8x1xf32>
    %216 = arith.mulf %215, %51 : tensor<8x1xf32>
    %217 = arith.truncf %216 : tensor<8x1xf32> to tensor<8x1xf16>
    %218 = arith.extf %75 : tensor<8x1xf16> to tensor<8x1xf32>
    %219 = arith.mulf %155, %218 : tensor<8x1xf32>
    %220 = arith.truncf %219 : tensor<8x1xf32> to tensor<8x1xf16>
    %221 = arith.extf %220 : tensor<8x1xf16> to tensor<8x1xf32>
    %222 = arith.extf %169 : tensor<8x1xf16> to tensor<8x1xf32>
    %223 = arith.addf %221, %222 : tensor<8x1xf32>
    %224 = arith.truncf %223 : tensor<8x1xf32> to tensor<8x1xf16>
    %225 = arith.extf %110 : tensor<8x1xf16> to tensor<8x1xf32>
    %226 = arith.extf %118 : tensor<8x1xf16> to tensor<8x1xf32>
    %227 = arith.addf %225, %226 : tensor<8x1xf32>
    %228 = arith.truncf %227 : tensor<8x1xf32> to tensor<8x1xf16>
    gpu.barrier
    %229 = tt.splat %arg2 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %230 = tt.addptr %229, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    tt.store %230, %169, %7 : tensor<8x1x!tt.ptr<f16>>
    gpu.barrier
    %231 = tt.splat %arg3 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %232 = tt.addptr %231, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    tt.store %232, %211, %7 : tensor<8x1x!tt.ptr<f16>>
    %233 = arith.addi %14, %5 : tensor<8x1xi32>
    %234 = tt.splat %arg13 : i32 -> tensor<8x1xi32>
    %235 = arith.muli %234, %14 : tensor<8x1xi32>
    %236 = arith.addi %233, %235 : tensor<8x1xi32>
    %237 = tt.splat %arg14 : i32 -> tensor<8x1xi32>
    %238 = arith.muli %237, %14 : tensor<8x1xi32>
    %239 = arith.addi %236, %238 : tensor<8x1xi32>
    %240 = tt.splat %arg15 : i32 -> tensor<8x1xi32>
    %241 = arith.muli %240, %14 : tensor<8x1xi32>
    %242 = arith.addi %239, %241 : tensor<8x1xi32>
    %243 = tt.splat %arg16 : i32 -> tensor<8x1xi32>
    %244 = arith.muli %243, %14 : tensor<8x1xi32>
    %245 = arith.addi %242, %244 : tensor<8x1xi32>
    %246 = tt.splat %arg12 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %247 = tt.addptr %246, %245 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    tt.store %247, %214, %7 : tensor<8x1x!tt.ptr<f16>>
    gpu.barrier
    %248 = tt.splat %arg0 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %249 = tt.addptr %248, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    tt.store %249, %217, %7 : tensor<8x1x!tt.ptr<f16>>
    gpu.barrier
    %250 = tt.splat %arg1 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %251 = tt.addptr %250, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    tt.store %251, %224, %7 : tensor<8x1x!tt.ptr<f16>>
    gpu.barrier
    %252 = tt.splat %arg4 : !tt.ptr<f16> -> tensor<8x1x!tt.ptr<f16>>
    %253 = tt.addptr %252, %5 : tensor<8x1x!tt.ptr<f16>>, tensor<8x1xi32>
    tt.store %253, %228, %7 : tensor<8x1x!tt.ptr<f16>>
    tt.return
  }
}

{-#
  external_resources: {
    mlir_reproducer: {
      pipeline: "builtin.module(convert-triton-to-tritongpu{enable-source-remat=false num-ctas=1 num-warps=4 target=cuda:120 threads-per-warp=32}, tritongpu-coalesce, tritongpu-F32DotTC, triton-nvidia-gpu-plan-cta, tritongpu-remove-layout-conversions, tritongpu-optimize-thread-locality, tritongpu-accelerate-matmul, tritongpu-remove-layout-conversions, tritongpu-optimize-dot-operands{hoist-layout-conversion=true}, triton-nvidia-optimize-descriptor-encoding, triton-loop-aware-cse, tritongpu-fuse-nested-loops, canonicalize{  max-iterations=10 max-num-rewrites=-1 region-simplify=normal test-convergence=false top-down=true}, triton-licm, tritongpu-optimize-accumulator-init, tritongpu-hoist-tmem-alloc{hoist-out-of-if=false}, tritongpu-promote-lhs-to-tmem, tritongpu-assign-latencies{num-stages=3}, tritongpu-schedule-loops, tritongpu-automatic-warp-specialization{num-stages=3}, tritongpu-pipeline{dump-intermediate-steps=false num-stages=3}, tritongpu-combine-tensor-select-and-if, tritongpu-hoist-tmem-alloc{hoist-out-of-if=true}, triton-nvidia-gpu-remove-tmem-tokens, canonicalize{  max-iterations=10 max-num-rewrites=-1 region-simplify=normal test-convergence=false top-down=true}, triton-loop-aware-cse, tritongpu-prefetch, tritongpu-optimize-dot-operands{hoist-layout-conversion=true}, tritongpu-coalesce-async-copy, triton-nvidia-optimize-tmem-layouts, triton-nvidia-tma-lowering, tritongpu-remove-layout-conversions, triton-nvidia-interleave-tmem, tritongpu-reduce-data-duplication, tritongpu-reorder-instructions, triton-loop-aware-cse, symbol-dce, triton-nvidia-gpu-fence-insertion{compute-capability=120}, triton-nvidia-mma-lowering, sccp, cse, canonicalize{  max-iterations=10 max-num-rewrites=-1 region-simplify=normal test-convergence=false top-down=true})",
      disable_threading: false,
      verify_each: true
    }
  }
#-}
/home/jansel/pytorch/triton_issue163567_min.py:51:0: error: Failures have been detected while processing an MLIR pass pipeline
/home/jansel/pytorch/triton_issue163567_min.py:51:0: note: Pipeline failed while executing [`TritonGPURemoveLayoutConversions` on 'builtin.module' operation]: reproducer generated at `std::errs, please share the reproducer above with Triton project.`
Traceback (most recent call last):
  File "/home/jansel/pytorch/triton_issue163567_min.py", line 473, in <module>
    main()
  File "/home/jansel/pytorch/triton_issue163567_min.py", line 444, in main
    triton_red_fused_add_div_eq_exp_ge_log_logical_and_masked_fill_max_mul_neg_pow_scatter_sub_unsqueeze_where_zeros_1[
  File "/home/jansel/conda/envs/pytorch/lib/python3.12/site-packages/triton/runtime/jit.py", line 359, in <lambda>
    return lambda *args, **kwargs: self.run(grid=grid, warmup=False, *args, **kwargs)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/conda/envs/pytorch/lib/python3.12/site-packages/triton/runtime/jit.py", line 675, in run
    kernel = self._do_compile(key, signature, device, constexprs, options, attrs, warmup)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/conda/envs/pytorch/lib/python3.12/site-packages/triton/runtime/jit.py", line 803, in _do_compile
    kernel = self.compile(src, target=target, options=options.__dict__)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/conda/envs/pytorch/lib/python3.12/site-packages/triton/compiler/compiler.py", line 320, in compile
    next_module = compile_ir(module, metadata)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/conda/envs/pytorch/lib/python3.12/site-packages/triton/backends/nvidia/compiler.py", line 516, in <lambda>
    stages["ttgir"] = lambda src, metadata: self.make_ttgir(src, metadata, options, capability)
                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jansel/conda/envs/pytorch/lib/python3.12/site-packages/triton/backends/nvidia/compiler.py", line 319, in make_ttgir
    pm.run(mod, 'make_ttgir')
RuntimeError: PassManager::run failed
```

### Environment details

Triton: 1b27b93fa5a4efdfef285581df4524189da81bba
GPU: RTX 5090

--- a/lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp
+++ b/lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp
@@ -132,20 +132,30 @@ public:
   void hoistConvertOnTopOfExtOrBroadcast(ConvertLayoutOp convertOp);
   void hoistConvertIntoConditionals();
   void hoistConvertIntoConditionals(ConvertLayoutOp convertOp);
-  void rewriteSlice(SetVector<Value> &slice, DenseMap<Value, Attribute> &layout,
-                    ConvertLayoutOp convertOp, IRMapping &mapping);
-  void rewriteSlice(SetVector<Value> &slice, DenseMap<Value, Attribute> &layout,
-                    ConvertLayoutOp convertOp);
+  void rewriteSlice(
+      SetVector<Value> &slice, DenseMap<Value, Attribute> &layout,
+      const DenseMap<std::pair<Value, Attribute>, Value> &existingRemats,
+      ConvertLayoutOp convertOp, IRMapping &mapping);
+  void rewriteSlice(
+      SetVector<Value> &slice, DenseMap<Value, Attribute> &layout,
+      const DenseMap<std::pair<Value, Attribute>, Value> &existingRemats,
+      ConvertLayoutOp convertOp);
 
-  LogicalResult
-  getConvertBackwardSlice(OpOperand &root, Attribute rootEncoding,
-                          SetVector<Value> &slice,
-                          DenseMap<Value, Attribute> &layout,
-                          std::function<bool(Operation *)> stopPropagation);
+  /// Invokes the utility function getConvertBackwardSlice with a callback for
+  /// checking whether a rematerialization for a particular value already
+  /// exists. Any value that has an existing rematerialization for all of its
+  /// uses will have that rematerialization inserted in \p existingRemats, and
+  /// will not have its operands traversed for inclusion in \p slice.
+  LogicalResult getConvertBackwardSlice(
+      OpOperand &root, Attribute rootEncoding, SetVector<Value> &slice,
+      DenseMap<Value, Attribute> &layout,
+      DenseMap<std::pair<Value, Attribute>, Value> &existingRemats,
+      std::function<bool(Operation *)> stopPropagation);
 
   LogicalResult getRematerializableSlice(
       OpOperand &root, Attribute rootEncoding, SetVector<Value> &slice,
       DenseMap<Value, Attribute> &layout,
+      DenseMap<std::pair<Value, Attribute>, Value> &existingRemats,
       std::function<bool(Operation *)> stopPropagation = nullptr);
 
 private:
@@ -670,10 +680,10 @@ void LayoutRematerialization::updateRematMapping(
   }
 }
 
-void LayoutRematerialization::rewriteSlice(SetVector<Value> &slice,
-                                           DenseMap<Value, Attribute> &layout,
-                                           ConvertLayoutOp convertOp,
-                                           IRMapping &mapping) {
+void LayoutRematerialization::rewriteSlice(
+    SetVector<Value> &slice, DenseMap<Value, Attribute> &layout,
+    const DenseMap<std::pair<Value, Attribute>, Value> &existingRemats,
+    ConvertLayoutOp convertOp, IRMapping &mapping) {
   SetVector<Operation *> opsToRewrite;
   // Keep track of yield operands that need to be duplicated.
   DenseMap<Operation *, SmallVector<int>> yieldOperandsMap;
@@ -684,8 +694,10 @@ void LayoutRematerialization::rewriteSlice(SetVector<Value> &slice,
   for (Value v : slice) {
     auto layoutIt = layout.find(v);
     assert(layoutIt != layout.end());
-    // If we already have a remat value for this value, use it.
-    if (Value remat = getRematValue(v, layoutIt->second)) {
+    // If we found a valid rematerialization for this value while constructing
+    // the slice, use that.
+    if (Value remat = existingRemats.lookup({v, layoutIt->second})) {
+      assert(getRematValue(v, layoutIt->second) == remat && "remat mismatch");
       mapping.map(v, remat);
       valuesWithExistingRemat.insert(v);
       continue;
@@ -836,20 +848,20 @@ void LayoutRematerialization::rewriteSlice(SetVector<Value> &slice,
     opToDelete.insert(op);
 }
 
-void LayoutRematerialization::rewriteSlice(SetVector<Value> &slice,
-                                           DenseMap<Value, Attribute> &layout,
-                                           ConvertLayoutOp convertOp) {
+void LayoutRematerialization::rewriteSlice(
+    SetVector<Value> &slice, DenseMap<Value, Attribute> &layout,
+    const DenseMap<std::pair<Value, Attribute>, Value> &existingRemats,
+    ConvertLayoutOp convertOp) {
   IRMapping mapping;
-  rewriteSlice(slice, layout, convertOp, mapping);
+  rewriteSlice(slice, layout, existingRemats, convertOp, mapping);
 }
 
 LogicalResult LayoutRematerialization::getConvertBackwardSlice(
     OpOperand &root, Attribute rootEncoding, SetVector<Value> &slice,
     DenseMap<Value, Attribute> &layout,
+    DenseMap<std::pair<Value, Attribute>, Value> &existingRemats,
     std::function<bool(Operation *)> stopPropagation) {
-  // Allow re-using existing conversions for a value. Check dominance of any
-  // reusable materializations against the root value. This is sufficient
-  // because the conversions are processed in post-order.
+  // Allow re-using existing conversions for a value if it dominates the use.
   auto getExistingConversion = [&](OpOperand &value, Attribute encoding) {
     Value remat = getRematValue(value.get(), encoding);
     if (!remat)
@@ -858,6 +870,7 @@ LogicalResult LayoutRematerialization::getConvertBackwardSlice(
     // dominates the current use of value.
     Operation *user = value.getOwner();
     if (domInfo.properlyDominates(remat, user)) {
+      existingRemats.try_emplace({value.get(), encoding}, remat);
       return remat;
     }
     // FIXME: If the current user is a conversion, then we know it will become
@@ -871,6 +884,10 @@ LogicalResult LayoutRematerialization::getConvertBackwardSlice(
     //   }
     //   return remat;
     // }
+
+    // There is an existing rematerialization, but it doesn't dominate all the
+    // uses we care about, so ensure it isn't used.
+    existingRemats[{value.get(), encoding}] = Value();
     return Value();
   };
 
@@ -881,13 +898,15 @@ LogicalResult LayoutRematerialization::getConvertBackwardSlice(
 LogicalResult LayoutRematerialization::getRematerializableSlice(
     OpOperand &root, Attribute rootEncoding, SetVector<Value> &sliceArg,
     DenseMap<Value, Attribute> &layoutArg,
+    DenseMap<std::pair<Value, Attribute>, Value> &existingRematsArg,
     std::function<bool(Operation *)> stopPropagation) {
   // Operate on copies of the input, we do not want to modify them unless we
   // have succeeded.
   auto slice = sliceArg;
   auto layout = layoutArg;
-  LogicalResult result = getConvertBackwardSlice(root, rootEncoding, slice,
-                                                 layout, stopPropagation);
+  auto existingRemats = existingRematsArg;
+  LogicalResult result = getConvertBackwardSlice(
+      root, rootEncoding, slice, layout, existingRemats, stopPropagation);
   if (result.failed() || slice.empty())
     return failure();
 
@@ -900,6 +919,7 @@ LogicalResult LayoutRematerialization::getRematerializableSlice(
   }
   sliceArg = std::move(slice);
   layoutArg = std::move(layout);
+  existingRematsArg = std::move(existingRemats);
   return success();
 }
 
@@ -1125,8 +1145,10 @@ void LayoutRematerialization::backwardRematerialization(
   // rematerialized.
   SetVector<Value> slice;
   DenseMap<Value, Attribute> layout;
+  DenseMap<std::pair<Value, Attribute>, Value> existingRemats;
   LogicalResult result = getRematerializableSlice(
-      convertOp.getSrcMutable(), targetType.getEncoding(), slice, layout);
+      convertOp.getSrcMutable(), targetType.getEncoding(), slice, layout,
+      existingRemats);
   if (result.failed()) {
     LDBG("  getRematerializableSlice failed");
     return;
@@ -1145,7 +1167,7 @@ void LayoutRematerialization::backwardRematerialization(
   });
 
   // 3. Rewrite the slice.
-  rewriteSlice(slice, layout, convertOp);
+  rewriteSlice(slice, layout, existingRemats, convertOp);
 }
 
 void LayoutRematerialization::hoistConvertDotOperand() {
@@ -1217,9 +1239,11 @@ void LayoutRematerialization::hoistConvertDotOperand(
 
   SetVector<Value> slice;
   DenseMap<Value, Attribute> layout;
+  DenseMap<std::pair<Value, Attribute>, Value> existingRemats;
   // Set-up the conversion "cache"
   LogicalResult result = getConvertBackwardSlice(
-      convertOp.getSrcMutable(), targetType.getEncoding(), slice, layout, stop);
+      convertOp.getSrcMutable(), targetType.getEncoding(), slice, layout,
+      existingRemats, stop);
   if (result.failed())
     return;
 
@@ -1269,7 +1293,7 @@ void LayoutRematerialization::hoistConvertDotOperand(
       DBGS() << "    " << v << '\n';
   });
 
-  rewriteSlice(innerSlice, layout, convertOp, mapping);
+  rewriteSlice(innerSlice, layout, existingRemats, convertOp, mapping);
 }
 
 // For convert left we try to hoist them above type extension to reduce the cost
@@ -1296,9 +1320,10 @@ void LayoutRematerialization::hoistConvertOnTopOfExtOrBroadcast(
   // 1. Take a backward slice of all the tensor dependencies.
   SetVector<Value> slice;
   DenseMap<Value, Attribute> layout;
+  DenseMap<std::pair<Value, Attribute>, Value> existingRemats;
   LogicalResult result = getRematerializableSlice(
       convertOp.getSrcMutable(), targetType.getEncoding(), slice, layout,
-      isExtOrBroadcastOp);
+      existingRemats, isExtOrBroadcastOp);
   if (result.failed())
     return;
 
@@ -1317,7 +1342,7 @@ void LayoutRematerialization::hoistConvertOnTopOfExtOrBroadcast(
     // If we can rematerialize the rest of the ext slice we can ignore this ext
     // as it won't need a convert.
     if (succeeded(getRematerializableSlice(op->getOpOperand(0), srcEncoding,
-                                           slice, layout)))
+                                           slice, layout, existingRemats)))
       continue;
 
     // Only apply it if there is a single ext op otherwise we would have to
@@ -1354,7 +1379,7 @@ void LayoutRematerialization::hoistConvertOnTopOfExtOrBroadcast(
   mapping.map(extOrBroadcastOp->getResult(0), newExtOrBroadcast->getResult(0));
   slice.remove(extOrBroadcastOp->getResult(0));
   // 3. Rewrite the slice.
-  rewriteSlice(slice, layout, convertOp, mapping);
+  rewriteSlice(slice, layout, existingRemats, convertOp, mapping);
 }
 
 void LayoutRematerialization::hoistConvertIntoConditionals(
@@ -1363,10 +1388,11 @@ void LayoutRematerialization::hoistConvertIntoConditionals(
   // stopping at conditionals. This subslice is used to initialize the analysis.
   SetVector<Value> slice;
   DenseMap<Value, Attribute> layout;
+  DenseMap<std::pair<Value, Attribute>, Value> existingRemats;
   auto isIfOp = [](Operation *op) { return isa<scf::IfOp>(op); };
   if (failed(getRematerializableSlice(convertOp.getSrcMutable(),
                                       convertOp.getType().getEncoding(), slice,
-                                      layout, isIfOp)))
+                                      layout, existingRemats, isIfOp)))
     return;
 
   // These are the conditional edges above which conversions should be hoisted.
@@ -1400,17 +1426,19 @@ void LayoutRematerialization::hoistConvertIntoConditionals(
 
     auto newSlice = slice;
     auto newLayout = layout;
+    auto newExistingRemats = existingRemats;
 
     LogicalResult thenResult = getRematerializableSlice(
-        thenRes, rootLayout, newSlice, newLayout, isIfOp);
+        thenRes, rootLayout, newSlice, newLayout, newExistingRemats, isIfOp);
     LogicalResult elseResult = getRematerializableSlice(
-        elseRes, rootLayout, newSlice, newLayout, isIfOp);
+        elseRes, rootLayout, newSlice, newLayout, newExistingRemats, isIfOp);
 
     // If propagation across both edges of this conditional succeeded, then we
     // don't need to hoist across it. Merge into the current slice.
     if (succeeded(thenResult) && succeeded(elseResult)) {
       slice = std::move(newSlice);
       layout = std::move(newLayout);
+      existingRemats = std::move(newExistingRemats);
       continue;
     }
 
@@ -1431,6 +1459,7 @@ void LayoutRematerialization::hoistConvertIntoConditionals(
 
     slice = std::move(newSlice);
     layout = std::move(newLayout);
+    existingRemats = std::move(newExistingRemats);
     // The layout conversion can be rematerialized along one edge but not the
     // other. We can hoist the conversion into the other branch. Push this
     // into the subslice list for analysis.
@@ -1465,7 +1494,7 @@ void LayoutRematerialization::hoistConvertIntoConditionals(
     OpBuilder b(edge->getOwner());
     hoistRemat(b, edge->get(), layout.at(result));
   }
-  rewriteSlice(slice, layout, convertOp, mapping);
+  rewriteSlice(slice, layout, existingRemats, convertOp, mapping);
 }
 
 bool backwardRematerialization(ModuleOp module) {

--- a/test/TritonGPU/combine.mlir
+++ b/test/TritonGPU/combine.mlir
@@ -4056,3 +4056,75 @@ module attributes {"ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 4 : i32, "ttg.thr
     tt.return %3 : tensor<32xf32, #ttg.slice<{dim = 0, parent = #ttg.slice<{dim = 0, parent = #blocked1}>}>>
   }
 }
+
+// -----
+
+// There was previously a bug where one of the layout conversions would be
+// incorrectly reused during backward rematerialization as an operand to an
+// instruction that preceded it.
+#blocked = #ttg.blocked<{sizePerThread = [2], threadsPerWarp = [32], warpsPerCTA = [4], order = [0]}>
+#blocked1 = #ttg.blocked<{sizePerThread = [1], threadsPerWarp = [32], warpsPerCTA = [4], order = [0]}>
+module attributes {"ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 4 : i32, ttg.target = "cuda:100", "ttg.threads-per-warp" = 32 : i32} {
+  tt.func public @kernel(%arg0: !tt.ptr<f32>) -> (tensor<8xf32, #blocked>, tensor<8xf32, #blocked>, tensor<8xf32, #blocked>) attributes {noinline = false} {
+    %0 = tt.make_range {end = 8 : i32, start = 0 : i32} : tensor<8xi32, #blocked1>
+    %1 = tt.splat %arg0 : !tt.ptr<f32> -> tensor<8x!tt.ptr<f32>, #blocked1>
+    %2 = tt.addptr %1, %0 : tensor<8x!tt.ptr<f32>, #blocked1>, tensor<8xi32, #blocked1>
+    %3 = tt.load %2 : tensor<8x!tt.ptr<f32>, #blocked1>
+    %4 = math.exp %3 : tensor<8xf32, #blocked1>
+    %5 = math.exp %4 : tensor<8xf32, #blocked1>
+    %6 = math.exp %5 : tensor<8xf32, #blocked1>
+    %7 = math.exp %6 : tensor<8xf32, #blocked1>
+    %8 = math.exp %7 : tensor<8xf32, #blocked1>
+    %9 = math.exp %8 : tensor<8xf32, #blocked1>
+    %10 = math.exp %9 : tensor<8xf32, #blocked1>
+    %11 = math.exp %10 : tensor<8xf32, #blocked1>
+    %12 = math.exp %11 : tensor<8xf32, #blocked1>
+    %13 = math.exp %12 : tensor<8xf32, #blocked1>
+    %14 = math.exp %13 : tensor<8xf32, #blocked1>
+    %15 = math.exp %14 : tensor<8xf32, #blocked1>
+    %16 = math.exp %15 : tensor<8xf32, #blocked1>
+    %17 = math.exp %16 : tensor<8xf32, #blocked1>
+    %18 = math.exp %17 : tensor<8xf32, #blocked1>
+    %19 = math.exp %18 : tensor<8xf32, #blocked1>
+    %20 = math.exp %19 : tensor<8xf32, #blocked1>
+    // %21 is too expensive to rematerialize, so we just record a mapping
+    // %19 -> %21 for future rematerializations.
+    %21 = ttg.convert_layout %19 : tensor<8xf32, #blocked1> -> tensor<8xf32, #blocked>
+    // %22 is just below the cost threshold, so we rematerialize the whole chain ending in %18.
+    %22 = ttg.convert_layout %18 : tensor<8xf32, #blocked1> -> tensor<8xf32, #blocked>
+    // Now that %18 is rematerialized in blocked1, the chain ending %20 is cheap
+    // enough to rematerialize. However, when rematerializing %19 as part of
+    // this chain, we must not consider %21, as it does not dominate %20.
+    %23 = ttg.convert_layout %20 : tensor<8xf32, #blocked1> -> tensor<8xf32, #blocked>
+    tt.return %21, %22, %23 : tensor<8xf32, #blocked>, tensor<8xf32, #blocked>, tensor<8xf32, #blocked>
+  }
+}
+
+// CHECK-NOT: ttg.convert_layout
+
+// -----
+
+// Test that we correctly process existing rematerializations when hoisting
+// converts on top of ext/broadcast ops.
+
+#blocked = #ttg.blocked<{sizePerThread = [1, 8], threadsPerWarp = [4, 8], warpsPerCTA = [4, 1], order = [1, 0]}>
+#blocked1 = #ttg.blocked<{sizePerThread = [4, 4], threadsPerWarp = [2, 16], warpsPerCTA = [4, 1], order = [1, 0]}>
+module attributes {"ttg.num-warps" = 4 : i32} {
+  tt.func @test_hoist_cvt_existing_remat(%arg0: tensor<128x64xi32, #blocked>) -> (tensor<128x64xi64, #blocked1>, tensor<128x64xi64, #blocked1>) {
+    %0 = tt.make_range {end = 128 : i32, start = 0 : i32} : tensor<128xi32, #ttg.slice<{dim = 1, parent = #blocked}>>
+    %1 = tt.expand_dims %0 {axis = 1 : i32} : tensor<128xi32, #ttg.slice<{dim = 1, parent = #blocked}>> -> tensor<128x1xi32, #blocked>
+    // TODO: We should be able to reuse the first conversion so there is just 1.
+    // CHECK: ttg.convert_layout %arg0
+    // CHECK-NOT: ttg.convert_layout
+    %2 = arith.extsi %arg0 : tensor<128x64xi32, #blocked> to tensor<128x64xi64, #blocked>
+    %3 = arith.extsi %1 : tensor<128x1xi32, #blocked> to tensor<128x1xi64, #blocked>
+    %4 = tt.broadcast %3 : tensor<128x1xi64, #blocked> -> tensor<128x64xi64, #blocked>
+    %5 = arith.muli %2, %4 : tensor<128x64xi64, #blocked>
+    %6 = ttg.convert_layout %5 : tensor<128x64xi64, #blocked> -> tensor<128x64xi64, #blocked1>
+    %7 = arith.extsi %1 : tensor<128x1xi32, #blocked> to tensor<128x1xi64, #blocked>
+    %8 = tt.broadcast %7 : tensor<128x1xi64, #blocked> -> tensor<128x64xi64, #blocked>
+    %9 = arith.muli %2, %8 : tensor<128x64xi64, #blocked>
+    %10 = ttg.convert_layout %9 : tensor<128x64xi64, #blocked> -> tensor<128x64xi64, #blocked1>
+    tt.return %6, %10 : tensor<128x64xi64, #blocked1>, tensor<128x64xi64, #blocked1>
+  }
+}

